Plugins/NativeCapture/Source/NativeCapture.cpp

```diff
@@ -4,15 +4,123 @@
 
 #include <Babylon/JsRuntime.h>
 #include <GraphicsImpl.h>
+#include <FrameBuffer.h>
 
 #include <vector>
 
+namespace
+{
+    using FrameCallback = std::function<void(uint32_t width, uint32_t height, bgfx::TextureFormat::Enum format, bool yFlip, gsl::span<const uint8_t> data)>;
+    using FrameProviderCleanup = std::function<void()>;
+    using FrameProviderTicket = gsl::final_action<FrameProviderCleanup>;
+
+    FrameProviderTicket BeginFrameCapture(Babylon::Graphics::Impl& graphicsImpl, bgfx::FrameBufferHandle frameBufferHandle, FrameCallback callback)
+    {
+        class DefaultBufferFrameProvider final : public std::enable_shared_from_this<DefaultBufferFrameProvider>
+        {
+        public:
+            static FrameProviderTicket Create(Babylon::Graphics::Impl& graphicsImpl, FrameCallback callback)
+            {
+                std::shared_ptr<DefaultBufferFrameProvider> frameProvider{new DefaultBufferFrameProvider(graphicsImpl, std::move(callback))};
+                frameProvider->StartCapture();
+                return gsl::finally<FrameProviderCleanup>([frameProvider{std::move(frameProvider)}]() mutable {
+                    frameProvider->m_ticket.reset();
+                });
+            }
+
+        private:
+            DefaultBufferFrameProvider(Babylon::Graphics::Impl& graphicsImpl, FrameCallback callback)
+                : m_graphicsImpl{graphicsImpl}
+                , m_frameCallback{std::move(callback)}
+            {
+            }
+
+            void StartCapture()
+            {
+                m_ticket = std::make_unique<Babylon::Graphics::Impl::CaptureCallbackTicketT>(m_graphicsImpl.AddCaptureCallback([thisRef{shared_from_this()}](auto& data) {
+                    thisRef->m_frameCallback(data.Width, data.Height, data.Format, data.YFlip, {static_cast<const uint8_t*>(data.Data), static_cast<std::ptrdiff_t>(data.DataSize)});
+                }));
+            }
+
+        private:
+            Babylon::Graphics::Impl& m_graphicsImpl;
+            std::unique_ptr<Babylon::Graphics::Impl::CaptureCallbackTicketT> m_ticket{};
+            FrameCallback m_frameCallback{};
+        };
+
+        class OffScreenBufferFrameProvider final : public std::enable_shared_from_this<OffScreenBufferFrameProvider>
+        {
+        public:
+            static FrameProviderTicket Create(Babylon::Graphics::Impl& graphicsImpl, bgfx::FrameBufferHandle frameBufferHandle, FrameCallback callback)
+            {
+                std::shared_ptr<OffScreenBufferFrameProvider> frameProvider{new OffScreenBufferFrameProvider(graphicsImpl, frameBufferHandle, std::move(callback))};
+                frameProvider->ReadTextureAsync();
```

Comment:
Might be good to comment here saying this is starting the capture and this function will call itself.